README
NAME
    PYX::SGML::Tags - Processing PYX data or file and write as SGML via
    Tags.

SYNOPSIS
     use PYX::SGML::Tags;

     my $obj = PYX::SGML::Tags->new(%parameters);
     $obj->parse($pyx, $out);
     $obj->parse_file($input_file, $out);
     $obj->parse_handler($input_file_handler, $out);
     $obj->finalize;

METHODS
  "new"
     my $obj = PYX::SGML::Tags->new(%parameters);

    Constructor.

    *       "input_encoding"

            Input encoding for parse_file() and parse_handler() usage.

            Default value is 'utf-8'.

    *       "input_tags_item_callback"

            Input 'Tags' item callback. This callback is for Tags::Output::*
            constructor parameter 'input_tags_item_callback'.

            Default value is undef.

    *       "output_encoding"

            Output encoding.

            Default value is 'utf-8'.

    *       "tags"

            Tags object. Can be any of Tags::Output::* objects. Default
            value is "Tags::Output::Raw-"new('output_handler' => \*STDOUT)>.
            It's required.

    Returns instance of class.

  "parse"
     $obj->parse($pyx, $out);

    Parse PYX text or array of PYX text. Output is serialization to SGML by
    Tags::Output::* module. If $out not present, use 'output_handler'.

    Returns undef.

  "parse_file"
     $obj->parse_file($input_file, $out);

    Parse file with PYX data. $input_file file is decoded by
    'input_encoding'. Output is serialization to SGML. If $out not present,
    use 'output_handler'.

    Returns undef.

  "parse_handler"
     $obj->parse_handler($input_file_handler, $out);

    Parse PYX handler. $input_file_handler handler is decoded by
    'input_encoding'. Output is serialization to SGML. If $out not present,
    use 'output_handler'.

    Returns undef.

  "finalize"
     $obj->finalize;

     Finalize opened tags, if exists.
     Returns undef.

ERRORS
     new():
             Bad 'Tags::Output::*' object.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

     parse():
             From PYX::Parser::parse():
                     Bad PYX line '%s'.
             From Tags::Output::Raw::flush():
                     Cannot write to output handler.

     parse_file():
             From PYX::Parser::parse_file():
                     Bad PYX line '%s'.
                     No input handler.
             From Tags::Output::Raw::flush():
                     Cannot write to output handler.

     parse_handler():
             From PYX::Parser::parse_handler():
                     Bad PYX line '%s'.
                     No input handler.
             From Tags::Output::Raw::flush():
                     Cannot write to output handler.

EXAMPLE1
     use strict;
     use warnings;

     use PYX::SGML::Tags;

     # Input.
     my $pyx = <<'END';
     (element
     -data
     )element
     END

     # Object.
     my $obj = PYX::SGML::Tags->new;

     # Process.
     $obj->parse($pyx);
     print "\n";

     # Output:
     # <element>data</element>

EXAMPLE2
     use strict;
     use warnings;

     use PYX::SGML::Tags;
     use Tags::Output::Indent;

     # Input.
     my $pyx = <<'END';
     (element
     -data
     )element
     END

     # Object.
     my $obj = PYX::SGML::Tags->new(
             'tags' => Tags::Output::Indent->new(
                     'output_handler' => \*STDOUT,
             ),
     );

     # Process.
     $obj->parse($pyx);
     print "\n";

     # Output:
     # <element>data</element>

EXAMPLE3
     use strict;
     use warnings;

     use PYX::SGML::Tags;
     use Tags::Output::Indent;

     # Input.
     my $pyx = <<'END';
     (element
     -data
     )element
     END

     # Object.
     my $obj = PYX::SGML::Tags->new(
             'input_tags_item_callback' => sub {
                     my $tags_ar = shift;
                     print '[ '.$tags_ar->[0].' ]'."\n";
                     return;
             },
     );

     # Process.
     $obj->parse($pyx);
     print "\n";

     # Output:
     # [ b ]
     # [ d ]
     # [ e ]
     # <element>data</element>

DEPENDENCIES
    Class::Utils, Error::Pure, PYX::Parser, PYX::Utils, Tags::Output::Raw.

SEE ALSO
    Task::PYX
        Install the PYX modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/PYX-SGML-Tags>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2011-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.11



Changes
0.11

0.10 2023-10-19T10:46:09+02:00
 - Fix typo in doc.
 - Improve constructor doc.
 - Update Module::Install to 1.21 version.
 - Update copyright years.

0.09 2022-11-02T23:25:09+01:00
 - Fix minimal versions of Tags::Output:: for tests. There was warning.
 - Rename examples to better names.
 - Update copyright years.

0.08 2021-02-22T17:51:23+01:00
 - Change api of parse(): Don't decode utf-8 string via 'input_encoding'.

0.07 2021-02-20T21:25:19+01:00
 - Add 'output_encoding' parameter.
 - Rewrite implicit encode() to PYX::Utils::encode().
 - Update copyright years.

0.06 2020-10-14T09:13:41+02:00
 - Add tests for finalize().
 - Change bugtracker to github.
 - Fix tests for situation when Tags::Output::LibXML not present.
 - Rewrite doc to use =head2 for methods.

0.05 2020-10-14T01:03:12+02:00
 - Add 'input_tags_item_callback' parameter for e.g.validation.
 - Add tests for output via libxml.
 - Rename test files to better names (raw tests).
 - Rewrite to better default of 'tags' parameter.
 - Rewrite Encode to Unicode::UTF8.

0.04 2020-04-26T18:32:33+02:00
 - Add test comments.
 - Add test for utf-8 characters (in Czech).
 - Add Tags reseting to each test.
 - Test for element renamed to -xml version. Added sgml version.

0.03 2020-04-26T15:28:41+02:00
 - Add EXAMPLE2 to PYX::SGML::Tags.
 - Add tests for parse() method.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Rename tests to better names (tag → element).
 - Rename PYX data files to better names (tag → element).
 - Rename PYX data to better names (tag → element).
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update copyright years.
 - Update my name to reality.

0.02 2016-02-20T23:38:52+01:00
 - Add 'input_encoding' parameter.
 - Add REPOSITORY section to doc.

0.01 2016-02-17T10:48:26+01:00
 - First version.


Tags.pm
package PYX::SGML::Tags;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use PYX::Parser;
use PYX::Utils;
use Tags::Output::Raw;

our $VERSION = 0.11;

# Constructor.
sub new {
	my ($class, @params) = @_;

	my $self = bless {}, $class;

	# Input encoding.
	$self->{'input_encoding'} = 'utf-8';

	# Input 'Tags' item callback.
	$self->{'input_tags_item_callback'} = undef;

	# Output encoding.
	$self->{'output_encoding'} = 'utf-8';

	# Tags object.
	$self->{'tags'} = undef;

	# Process params.
	set_params($self, @params);

	if (! defined $self->{'tags'}) {
		$self->{'tags'} = Tags::Output::Raw->new(
			'input_tags_item_callback'
				=> $self->{'input_tags_item_callback'},
			'output_encoding' => $self->{'output_encoding'},
			'output_handler' => \*STDOUT,
		);
	}

	# Check for Tags::Output object.
	if (! $self->{'tags'}->isa('Tags::Output')) {
		err "Bad 'Tags::Output::*' object.";
	}

	# PYX::Parser object.
	$self->{'pyx_parser'} = PYX::Parser->new(
		'callbacks' => {
			'attribute' => \&_attribute,
			'comment' => \&_comment,
			'data' => \&_data,
			'end_element' => \&_end_element,
			'instruction' => \&_instruction,
			'start_element' => \&_start_element,
		},
		'input_encoding' => $self->{'input_encoding'},
		'non_parser_options' => {
			'tags' => $self->{'tags'},
		},
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;

	$self->{'pyx_parser'}->parse($pyx, $out);
	$self->{'tags'}->flush;

	return;
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file, $out) = @_;

	$self->{'pyx_parser'}->parse_file($file, $out);
	$self->{'tags'}->flush;

	return;
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;

	$self->{'pyx_parser'}->parse_handler($input_file_handler, $out);
	$self->{'tags'}->flush;

	return;
}

sub finalize {
	my $self = shift;

	$self->{'tags'}->finalize;

	return;
}

# Process start of element.
sub _start_element {
	my ($self, $elem) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['b', $elem]);

	return;
}

# Process end of element.
sub _end_element {
	my ($self, $elem) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['e', $elem]);

	return;
}

# Process data.
sub _data {
	my ($self, $data) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['d', PYX::Utils::encode($data)]);

	return;
}

# Process attribute.
sub _attribute {
	my ($self, $attr, $value) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['a', $attr, $value]);

	return;
}

# Process instruction tag.
sub _instruction {
	my ($self, $target, $code) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['i', $target, $code]);

	return;
}

# Process comments.
sub _comment {
	my ($self, $comment) = @_;

	my $tags = $self->{'non_parser_options'}->{'tags'};
	$tags->put(['c', PYX::Utils::encode($comment)]);

	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::SGML::Tags - Processing PYX data or file and write as SGML via Tags.

=head1 SYNOPSIS

 use PYX::SGML::Tags;

 my $obj = PYX::SGML::Tags->new(%parameters);
 $obj->parse($pyx, $out);
 $obj->parse_file($input_file, $out);
 $obj->parse_handler($input_file_handler, $out);
 $obj->finalize;

=head1 METHODS

=head2 C<new>

 my $obj = PYX::SGML::Tags->new(%parameters);

Constructor.

=over 8

=item * C<input_encoding>

Input encoding for parse_file() and parse_handler() usage.

Default value is 'utf-8'.

=item * C<input_tags_item_callback>

Input 'Tags' item callback.
This callback is for Tags::Output::* constructor parameter 'input_tags_item_callback'.

Default value is undef.

=item * C<output_encoding>

Output encoding.

Default value is 'utf-8'.

=item * C<tags>

Tags object.
Can be any of Tags::Output::* objects.
Default value is C<Tags::Output::Raw->new('output_handler' => \*STDOUT)>.
It's required.

=back

Returns instance of class.

=head2 C<parse>

 $obj->parse($pyx, $out);

Parse PYX text or array of PYX text.
Output is serialization to SGML by Tags::Output::* module.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_file>

 $obj->parse_file($input_file, $out);

Parse file with PYX data.
C<$input_file> file is decoded by 'input_encoding'.
Output is serialization to SGML.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_handler>

 $obj->parse_handler($input_file_handler, $out);

Parse PYX handler.
C<$input_file_handler> handler is decoded by 'input_encoding'.
Output is serialization to SGML.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<finalize>

 $obj->finalize;

 Finalize opened tags, if exists.
 Returns undef.

=head1 ERRORS

 new():
         Bad 'Tags::Output::*' object.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 parse():
         From PYX::Parser::parse():
                 Bad PYX line '%s'.
         From Tags::Output::Raw::flush():
                 Cannot write to output handler.

 parse_file():
         From PYX::Parser::parse_file():
                 Bad PYX line '%s'.
                 No input handler.
         From Tags::Output::Raw::flush():
                 Cannot write to output handler.

 parse_handler():
         From PYX::Parser::parse_handler():
                 Bad PYX line '%s'.
                 No input handler.
         From Tags::Output::Raw::flush():
                 Cannot write to output handler.

=head1 EXAMPLE1

=for comment filename=simple_element_raw.pl

 use strict;
 use warnings;

 use PYX::SGML::Tags;

 # Input.
 my $pyx = <<'END';
 (element
 -data
 )element
 END

 # Object.
 my $obj = PYX::SGML::Tags->new;

 # Process.
 $obj->parse($pyx);
 print "\n";

 # Output:
 # <element>data</element>

=head1 EXAMPLE2

=for comment filename=simple_element_indent.pl

 use strict;
 use warnings;

 use PYX::SGML::Tags;
 use Tags::Output::Indent;

 # Input.
 my $pyx = <<'END';
 (element
 -data
 )element
 END

 # Object.
 my $obj = PYX::SGML::Tags->new(
         'tags' => Tags::Output::Indent->new(
                 'output_handler' => \*STDOUT,
         ),
 );

 # Process.
 $obj->parse($pyx);
 print "\n";

 # Output:
 # <element>data</element>

=head1 EXAMPLE3

=for comment filename=simple_element_callback.pl

 use strict;
 use warnings;

 use PYX::SGML::Tags;
 use Tags::Output::Indent;

 # Input.
 my $pyx = <<'END';
 (element
 -data
 )element
 END

 # Object.
 my $obj = PYX::SGML::Tags->new(
         'input_tags_item_callback' => sub {
                 my $tags_ar = shift;
                 print '[ '.$tags_ar->[0].' ]'."\n";
                 return;
         },
 );

 # Process.
 $obj->parse($pyx);
 print "\n";

 # Output:
 # [ b ]
 # [ d ]
 # [ e ]
 # <element>data</element>

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<PYX::Parser>,
L<PYX::Utils>,
L<Tags::Output::Raw>.

=head1 SEE ALSO

=over

=item L<Task::PYX>

Install the PYX modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/PYX-SGML-Tags>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2011-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.11

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Processing PYX data or file and write as SGML via Tags.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'PYX-SGML-Tags';
readme_from 'Tags.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.05;
requires 'Error::Pure' => 0.15;
requires 'PYX::Parser' => 0.08;
requires 'PYX::Utils' => 0.01;
requires 'Tags::Output::Raw' => 0.11;
requires 'perl' => '5.6.2';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/PYX-SGML-Tags/issues',
resources 'homepage' => 'https://github.com/michal-josef-spacek/PYX-SGML-Tags';
resources 'repository' => 'git://github.com/michal-josef-spacek/PYX-SGML-Tags';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.15;
test_requires 'File::Object' => 0.07;
test_requires 'Tags::Output::Raw' => 0.13;
test_requires 'Tags::Output::Indent' => 0.09;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Unicode::UTF8' => 0;
tests_recursive;
version '0.11';

# Run.
WriteAll();
